### PR TITLE
Improve build performance

### LINKED ISSUE #297: Circle CI out of memory error

### PR DESCRIPTION
**Pull request by ecgreb**
### Overview

Improves build performance on Circle CI. Reduces build time by 1-2 min for dev builds and 4-5 min for full master builds (includes sample apps). Prevents OOM errors on Circle CI container.

### Proposed Changes

* Updates Gradle options to limit memory use and dump heap on OOM error.
* Runs unit tests against debug build only and reduces console output.
* Upgrades Gradle to version 3.3 and gradle-android-maven-plugin to version 1.5.
* Adds `--configure-on-demand` and `-PdisablePreDex` options to circle.yml.
* Runs tests for each module in parallel using 3 Circle CI containers instead of 1.
* Archives sample app APKs on Circle CI and deploys to AWS on release build only.
* Publishes documentation to Mapzen website on release build only.
* Installs `core` module to local .m2 repo before resolving dependencies.
* Disables Javadoc tasks for all modules. May need to be re-enabled for production AARs.

Fixes #297

### CODE REVIEW

**Comment by tallytalwar** [comment]
Starting reviewing this now.

File: circle.yml [hunk 1]
```diff
@@ -19,7 +19,7 @@ dependencies:
 
 test:
   override:
-    - ./gradlew clean :core:install :core:verify :mapzen-android-sdk:verify :mapzen-places-api:verify
+    - ./gradlew clean :core:install :core:verify :mapzen-android-sdk:verify :mapzen-places-api:verify --configure-on-demand
```
Comment (tallytalwar) [inline circle.yml]: 👍 Nicely documented in gradle docs (which sometimes are very convoluted, at least for me): https://docs.gradle.org/current/userguide/multi_project_builds.html#sec:configuration_on_demand

File: circle.yml [hunk 2]
```diff
@@ -19,7 +19,7 @@ dependencies:
 
 test:
   override:
-    - ./gradlew clean :core:install :core:verify :mapzen-android-sdk:verify :mapzen-places-api:verify --configure-on-demand
+    - ./gradlew clean :core:install :core:verify :mapzen-android-sdk:verify :mapzen-places-api:verify --configure-on-demand -PdisablePreDex
```
Comment (tallytalwar) [inline circle.yml]: The only information I found for this is its recommended to use `-PdisablePreDex` when doing a one time build (typical with CI env.). But no finer details behind the reasoning. Can you point me to a reference/explain? Thanks

File: scripts/publish-docs.sh [hunk 3]
```diff
@@ -0,0 +1,9 @@
+#!/usr/bin/env bash
+#
+# Triggers mapzen docs build to publish to https://mapzen.com/documentation/.
+
+if [[ ${PERFORM_RELEASE} ]]
```
Comment (tallytalwar) [inline scripts/publish-docs.sh]: naive question ... when does `PERFORM_RELEASE` get set?

File: circle.yml (hunk 2 above)
Comment (ecgreb) [inline circle.yml]: Using `-PdisablePreDex` was recommended in [this Circle CI issue](https://discuss.circleci.com/t/java-lang-outofmemoryerror-permgen-space-in-android-gradle-build/3652).

File: scripts/publish-docs.sh (hunk 3 above)
Comment (ecgreb) [inline scripts/publish-docs.sh]: Haha it does not yet! :) I will be adding a script to perform the release using Circle's API similar to [this one we use for Eraser Map](https://github.com/mapzen/eraser-map/blob/master/scripts/perform-release.sh) as part of https://github.com/mapzen/android/issues/279.

**Pull request merged**